shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1024,13 +1096,19 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       case pat if ctx.inPattern =>
         err(msg"Unsupported pattern shape${
           if (dbg) " ("+pat.getClass.toString+")" else ""}:", pat.toLoc)(raise)
-      case Lam(pat, body) if doGenLambdas =>
-        println(s"TYPING POLY LAM")
-        ctx.nest.poly { newCtx =>
+      case Lam(pat, body) if doGenLambdas || ctx.inQuote =>
```

Comment:
Any reason you insist on treating quoted lambdas differently than unquoted ones? Why do you want this case to always be used when the lambda is quoted? I talked about this in my previous comment: you're making the typing of lambdas needlessly inconsistent, AFAICT.

Reply:
To track the contextual types of a quoted term, there are two strategies:
- For each binding, we track free variables inside the binding's body and solve the constraint. Take `code"x => x + 1"` as an example, we record free variables in `x + 1` (i.e., a skolem `gx`) and create a new type variable `a`. Then we solve the constraint `gx <= gx \/ a`.
- We treat quoted bindings as normal ones but track the contextual types when there is an `Unquote`. Notice that if there is no `Unquote`, there is no unbound quoted variable. We've inserted `Unquote` for each binding body in new syntax (i.e., `` x `=> ... ``). For `code"x => x + 1"`, we can assert that the contextual type is `Bot`. We only need to consider situations where there is an `Unquote`, like `code"x => ${id(x)}"`.
The first one is the current version. It did bring inconsistency to the typing of lambdas, but it works to both syntaxes (`code"x => ..."` and `` x `=> ... ``, the later one is equivalent to `code"x => ${...}"`) the same way, is consistent to our formal method, and can avoid wrong levels of variables when handling `Unquote` via the nature of polymorphic lambdas. So I still think that the first solution is better.

Reply:
Do you understand why there are two lambda cases in the original code?

Reply:
To avoid unnecessary generalizations. So are you suggesting that we have
```scala
val newCtx = if (ctx.inQuote) ctx.enterQuotedScope else ctx.nest
// ...

if (ctx.inQuote) {
  val ctxTy = freshVar(noTyProv, N)(ctx.lvl)
  ctx.trackFVs(con(newCtx.getCtxTy, newCtx.getAllQuoteSkolemsWith(ctxTy), ctxTy)(ctx))
}
// ...
```
in both cases of `Lam`?

Reply:
Yes. What is it you don't understand in:

> Just add these lines to the two `Lam` cases below.

?

I find it quite concerning that you decided to use one case and disregard the other, with no apparent reason. The changes you make to this crucial part of the code (the type checker) need to be well reasoned and well motivated.